compiler/rustc_middle/src/mir/interpret/error.rs

```diff
@@ -319,110 +363,81 @@ pub enum UndefinedBehaviorInfo {
+                    span,
+                    alloc_id,
+                    alloc_size: alloc_size.bytes(),
+                    ptr_offset,
+                    ptr_size: ptr_size.bytes(),
+                    is_zero: ptr_size == Size::ZERO,
+                    msg,
+                }
             }
-            PointerOutOfBounds { alloc_id, alloc_size, ptr_offset, ptr_size, msg } => write!(
-                f,
-                "{msg}{alloc_id:?} has size {alloc_size}, so pointer to {ptr_size} byte{ptr_size_p} starting at offset {ptr_offset} is out-of-bounds",
-                alloc_size = alloc_size.bytes(),
-                ptr_size = ptr_size.bytes(),
-                ptr_size_p = pluralize!(ptr_size.bytes()),
-            ),
-            DanglingIntPointer(i, msg) => {
-                write!(
-                    f,
-                    "{msg}{pointer} is a dangling pointer (it has no provenance)",
-                    pointer = Pointer::<Option<AllocId>>::from_addr_invalid(*i),
-                )
+            // FIXME the use of to_string here prints "null pointer" when 0, but we want that to be translatable
+            DanglingIntPointer(pointer, msg) => I::DanglingIntPointer {
+                span,
+                pointer: Pointer::<Option<AllocId>>::from_addr_invalid(pointer).to_string(),
+                msg,
+            },
+            AlignmentCheckFailed { required, has } => {
+                I::AlignmentCheckFailed { span, required: required.bytes(), has: has.bytes() }
             }
-            AlignmentCheckFailed { required, has } => write!(
-                f,
-                "accessing memory with alignment {has}, but alignment {required} is required",
-                has = has.bytes(),
-                required = required.bytes()
-            ),
-            WriteToReadOnly(a) => write!(f, "writing to {a:?} which is read-only"),
-            DerefFunctionPointer(a) => write!(f, "accessing {a:?} which contains a function"),
-            DerefVTablePointer(a) => write!(f, "accessing {a:?} which contains a vtable"),
-            ValidationFailure { path: None, msg } => {
-                write!(f, "constructing invalid value: {msg}")
+            InvalidBool(val) => I::InvalidBool { span, val: format!("{val:02x}") },
+            InvalidChar(val) => I::InvalidChar { span, val: format!("{val:08x}") },
+            InvalidTag(val) => I::InvalidTag { span, val: format!("{val:x}") },
+            InvalidStr(err) => I::InvalidStr { span, err: err.to_string() },
+            InvalidUninitBytes(Some((alloc, info))) => {
+                I::InvalidUninitBytes { span, alloc, access: info.access, uninit: info.uninit }
             }
-            ValidationFailure { path: Some(path), msg } => {
-                write!(f, "constructing invalid value at {path}: {msg}")
+            ScalarSizeMismatch(self::ScalarSizeMismatch { target_size, data_size }) => {
+                I::ScalarSizeMismatch { span, target_size, data_size }
```

Comment:
There's a *ton* of redundancy and duplication in here now, as shown by these big `match`es which do nothing but translate from one enum to another. Is there any way to avoid/reduce that?

Reply:
This is kind of hard, since the span of the diagnostic is not known until we decide to emit it. (many uses use the stack trace to find out where to point at) And this would have to either be a manual impl doing all the `span_note` and `set_arg` or using a `derive` to generate that which requires translating enums.

Reply:
Somehow for the `Custom` variant this seems to work though? Or does that vie up on some of the compile-time checks or so?

Reply:
I think what I don't get is why this can't directly refer to `fluent::middle_scalar_size_mismatch` or something like that. There's an indirection going on with that `I::ScalarSizeMismatch` type. Is that related to the `span_note`/`set_arg` things you mentioned?

Reply:
Yeah, with translatable diagnostics these messages are supposed to exist as `struct`/`enum`s that derive the `IntoDiagnostic` trait. Manual implementations are still supposed to use `IntoDiagnostic` and those can use something like `err.span_note(span, fluent::middle_scalar_size_mismatch)`. One of my early approaches was to make `InterpError` return a `DiagnosticMessage` such that it could exist both as a label and as an error. But after that I looked at uses of InterpError in rustc and thought that its only use was as a span_label so I chose to use span_label.

That said, we can use some kind of a hack by inspecting the diagnostic after we insert the interperror label in miri so I'm not sure if we should revert back to message only

Reply:
In my view, long-term there should be two things defined for each `InterpError`: a 'title' and a 'label', with the intention that the title is rendered as the main error text and the label is rendered at the span.

I am not sure what is the best way to represent this. I had planned to just make this two separate functions that return a string each, but that was pre-translation obviously.

Reply:
It can be done, by making both the title and the label `DiagnosticMessage`s (two functions) and have two functions that add the required arguments (non-translatable parts such as type paths) for them respectively. That could be the followup work.

Reply:
I [created a thread](https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n/topic/.60derive.28Subdiagnostic.29.60.2C.20but.20span.20comes.20from.20somewhere.20else) about this on Zulip. I think the easiest way forward would be to not use these traits but use inherent methods that either return the message or add arguments for formatting.

Reply:
Yeah it is fine to do this in follow-up changes. I just mentioned it in case it would affect your design decisions now.